File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
// binary search

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, 1);
        for(int i=0; i<n; i++){
            int maxi = 0;
            for(int j=0; j<i; j++){
                if(nums[j]>=nums[i]) continue;
                maxi = max(maxi, dp[j]);
            }
            dp[i] = 1+maxi;
        }

        return *max_element(dp.begin(), dp.end());
    }
};

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.go
func lengthOfLIS(nums []int) int {
    if(len(nums) == 0){
        return 0
    }
    n := len(nums)
    dp := make([]int, n)

    for i := 0; i<n; i++ {
        maxi := 0
        for j:=0; j<i; j++ {
            if nums[j] >= nums[i]{
                continue
            }
            maxi = max(maxi, dp[j])
        }
        dp[i] = maxi+1
    }
    
    maxima := 1;
    for _, x := range dp{
        maxima = max(maxima, x)
    }
    return maxima
}